libdnf5/base/base.cpp

```diff
@@ -214,6 +228,105 @@ void Base::setup() {
         p_impl->config.get_system_cachedir_option().set(Option::Priority::INSTALLROOT, full_path.string());
     }
 
+    // have with-mounts, mount specials
+    if (installroot_path != "/" && with_mounts) {
+        std::vector<std::string> dirs = {"run", "proc", "sys", "dev", "var", "tmp"};
+        std::vector<std::string> tmpfsvols = {"run", "tmp", "dev"};
+        std::vector<std::string> hostdevs = {"null", "zero", "full", "urandom", "tty"};
+        auto tpath = installroot_path;
+        int fd = -1;
+
+        // create any required top level directories
+        for (auto dir : dirs) {
+            tpath = installroot_path / dir;
+            mode_t mode = S_IRWXU | S_IRGRP | S_IXGRP | S_IROTH | S_IXOTH;
+
+            if (dir == "tmp") {
+                // set 1777 for mode on /tmp
+                mode = S_ISVTX | S_IRWXU | S_IRWXG | S_IRWXO;
+            }
+
+            if (libdnf5::utils::fs::mkdirp(tpath.c_str(), mode)) {
+                libdnf_throw_assertion("mkdir(2) failure: {}", strerror(errno));
+            }
+        }
+
+        // mount tmpfs filesystems
+        for (auto dir : tmpfsvols) {
+            tpath = installroot_path / dir;
+
+            if (libdnf5::utils::proc::call(PATH_TO_MOUNT, {"-t", "tmpfs", "tmpfs", tpath.c_str()})) {
+                libdnf_throw_assertion("libdnf5::utils::proc::call() failure: {}", strerror(errno));
+            }
+
+            if (dir == "tmp") {
+                if (chmod(tpath.c_str(), S_ISVTX | S_IRWXU | S_IRWXG | S_IRWXO)) {
+                    libdnf_throw_assertion("chmod(2) failure: {}", strerror(errno));
+                }
+            }
+        }
+
+        // some things (e.g., authselect) rely on this mount
+        tpath = installroot_path / "run/ostree-booted";
+        fd = open(tpath.c_str(), O_WRONLY | O_CREAT, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH);
+
+        if (fd == -1) {
+            warn("open");
+        }
+
+        if (close(fd) < 0) {
+            warn("close");
+        }
+
+        // bind mount /proc
+        tpath = installroot_path / "proc";
+
+        if (libdnf5::utils::proc::call(PATH_TO_MOUNT, {"--bind", "/proc", tpath.c_str()})) {
```

Comment:
Each of the calls to `mount` will want `--options-source=disable` and I think `--no-mtab`.